### PR TITLE
Add view once viewing

### PR DESCRIPTION
**Pull request by IsaacHatton**
This adds view once viewing functionality, note that none of the security/only view once functionality is present. 
Also, the image/video cannot be enlarged into the main viewer like other images/videos.
image

### CODE REVIEW

**Comment by IsaacHatton** [comment]
This fixes issue #128

File: core/interception.js [hunk 1]
```diff
@@ -454,6 +454,73 @@ NodeHandler.onMessageNodeReceived = async function(currentNode, messageNodes, is
         if (!isAllowed) break;
     }
 
+    if(encNodes[0].viewOnceMessageV2 !== null)
```
Comment (tomer8007) [inline core/interception.js:457]: Let's split this logic into a separate function, in a similar way to how message revoke is handled.

File: core/interception.js [hunk 2]
```diff
@@ -454,6 +454,73 @@ NodeHandler.onMessageNodeReceived = async function(currentNode, messageNodes, is
...
+
+        const decryptedData = await WhatsAppAPI.downloadManager.downloadAndMaybeDecrypt({ directPath: retrievedMsg.directPath, 
+            encFilehash: encodedencFileHash, filehash: encodedfileSha256, mediaKey: mediaKeyEncoded, 
+            type: type, signal: (new AbortController).signal });
+
+        body = arrayBufferToBase64(decryptedData);
+        dataURI = "data:" + retrievedMsg.mimetype + ";base64," + body
+        // store in indexedDB called "view-once" messageID and dataURI 
+        var viewOnceDBOpenRequest = indexedDB.open("viewOnce", 2);
+        viewOnceDBOpenRequest.onupgradeneeded = function (event)
```
Comment (tomer8007) [inline core/interception.js]: Here you actually save the view once message to the DB, not in `saveViewOnceToIndexedDBIfNecessary`

File: core/ui.js [hunk 3]
```diff
@@ -604,11 +604,74 @@ function onNewMessageNodeAdded(messageNode)
...
+    // check if there are any view-once messages for this message ID
+    // open viewOnce indexedDB
+    saveViewOnceToIndexedDBIfNecessary(msgID, messageNode);
+
     restoreDeletedMessageIfNeeded(messageNode, msgID);
 
     markMessageNodeDeviceIfPossible(messageNode, msgID);
 }
 
+function saveViewOnceToIndexedDBIfNecessary(msgID, messageNode)
```
Comment (tomer8007) [inline core/ui.js:616]: But this function does not save the view once message to the DB, it restores it

File: core/ui.js [hunk 4]
```diff
@@ -606,14 +606,14 @@ function onNewMessageNodeAdded(messageNode)
 
     // check if there are any view-once messages for this message ID
     // open viewOnce indexedDB
-    saveViewOnceToIndexedDBIfNecessary(msgID, messageNode);
+    checkIfMessageIDIsInViewOnceDB(msgID, messageNode);
```
Comment (tomer8007) [inline core/ui.js:609]: Let's call this `restoreViewOnceMessageIfNeeded`, like `restoreDeletedMessageIfNeeded`

File: core/ui.js [hunk 5]
```diff
@@ -606,14 +606,14 @@ function onNewMessageNodeAdded(messageNode)
 
     // check if there are any view-once messages for this message ID
     // open viewOnce indexedDB
-    checkIfMessageIDIsInViewOnceDB(msgID, messageNode);
+    restoreDeletedMessageIfNeeded(msgID, messageNode);
```
Comment (tomer8007) [inline core/ui.js:609]: That's the wrong name

File: core/ui.js (hunk 4 above)
Comment (IsaacHatton) [reply core/ui.js]: Sorry I somehow misread your comment.

File: core/interception.js [hunk 6]
```diff
@@ -454,6 +454,8 @@ NodeHandler.onMessageNodeReceived = async function(currentNode, messageNodes, is
         if (!isAllowed) break;
     }
 
+    await interceptViewOnceMessages(encNodes, messageId);
```
Comment (tomer8007) [inline core/interception.js:457]: This should be inside `onE2EMessageNodeReceived`, like in `checkForMessageDeletionNode`

File: core/ui.js [hunk 7]
```diff
@@ -604,11 +604,74 @@ function onNewMessageNodeAdded(messageNode)
...
+    // check if there are any view-once messages for this message ID
+    // open viewOnce indexedDB
+    restoreViewOnceMessageIfNeeded(msgID, messageNode);
+
     restoreDeletedMessageIfNeeded(messageNode, msgID);
 
     markMessageNodeDeviceIfPossible(messageNode, msgID);
 }
 
+function restoreViewOnceMessageIfNeeded(msgID, messageNode)
```
Comment (tomer8007) [inline core/ui.js:616]: Let's swap the parameters order here to be similar to `restoreDeletedMessageIfNeeded` and `markMessageNodeDeviceIfPossible`

**Comment by tomer8007** [comment]
Do you want to also mark the restored message in red, similarly to deleted messages?

**Comment by IsaacHatton** [comment]
I could as long as you think the view once logo will be enough for people to understand that it's a view once rather than a deleted normal message?

**Comment by tomer8007** [comment]
Can you post a screenshot of how the view once message usually looks like after you've already viewed it?

**Comment by IsaacHatton** [comment]
It doesn't change how it looks, my code never deletes the image once its viewed, the images are viewable unlimited times

**Comment by tomer8007** [comment]
I meant how does the user see a view-once message he's already viewed, without any modifications (the natural way)

**Comment by IsaacHatton** [comment]
It doesn't change how it looks, my code never deletes the image once its viewed, the images are viewable unlimited times.

Do you want it to enforce view once?

**Comment by tomer8007** [comment]
No, I was just wondering if it'll be clear to the user that the extension is enabling him to view the message twice even though it's a view-once message

**Comment by IsaacHatton** [comment]
IIRC the web has no change once a view once is opened on mobile, it always displays the default text about only viewable on mobile.

**Comment by tomer8007** [comment]
Oh I understand. So with the extension on, you will see view-once messages in the web, and you could view them unlimited times. Correct?

**Comment by IsaacHatton** [comment]
Yes, this is the functionality without this:
image

**Comment by tomer8007** [comment]
Finally, is everything working?

**Comment by IsaacHatton** [comment]
As far as I can tell, but I'm not confident with how the IndexedDB will behave when it has to create the database as I havent used indexedDB before.

**Comment by tomer8007** [comment]
I think it'll work, but you can just delete the IndexedDB / uninstall the extension and install it again to see.

**Comment by IsaacHatton** [comment]
I did that but my concern is that what happens if it gets created by the injection script as in my test account it always is created by UI because there are view once messages in the chat history.

**Comment by tomer8007** [comment]
I don't understand why you can't check the injection script flow.

**Comment by IsaacHatton** [comment]
Apologies, my earlier comment was incorrect, I just tested the injection script flow by sending myself before opening any chats, and it worked fine.

I forgot that the UI script would only run once a chat was opened, as there aren't any messages to check on the first screen (the one that recommends downloading the desktop app).

**Comment by tomer8007** [comment]
OK, Thank you for this pull request, your time is appreciated

**Comment by IsaacHatton** [comment]
Did you want the text added, if not don't worry:
image

**Comment by tomer8007** [comment]
I think it's a good text, you can have another pull request for it

**Pull request merged**